solidity/test/TokenTimeWeightedAverage.js

```diff
@@ -0,0 +1,502 @@
+const { accounts, defaultSender, contract, web3 } = require('@openzeppelin/test-environment');
+const { expectRevert, expectEvent, constants, BN, time } = require('@openzeppelin/test-helpers');
+const { expect } = require('../../chai-local');
+
+const Decimal = require('decimal.js');
+
+const { ZERO_ADDRESS } = constants;
+const { latest, duration } = time;
+
+const TokenTimeWeightedAverage = contract.fromArtifact('TestTokenTimeWeightedAverage');
+
+const ROLE_OWNER = web3.utils.keccak256('ROLE_OWNER');
+const ROLE_SEEDER = web3.utils.keccak256('ROLE_SEEDER');
+
+const initTWA = (start) => {
+    return {
+        firstSampleTime: Decimal(start.toString()),
+        lastSampleTime: Decimal(start.toString()),
+        accumulators: {
+            [start]: Decimal(0)
+        },
+        lastAccumulator: Decimal(0),
+        lastAccumulatorTime: Decimal(0)
+    };
+};
+
+const addTWASample = (acc, n, d, time) => {
+    const sampleTime = time.toNumber();
+    const value = Decimal(n.toString()).div(Decimal(d.toString()));
+    const { accumulators, lastSampleTime } = acc;
+
+    acc.lastAccumulator = accumulators[lastSampleTime];
+    acc.lastAccumulatorTime = sampleTime;
+
+    accumulators[sampleTime] = acc.lastAccumulator.add(Decimal(sampleTime).sub(lastSampleTime).mul(value));
+    acc.lastSampleTime = sampleTime;
+};
+
+const revertTWALastSample = (acc) => {
+    const { lastAccumulator, lastAccumulatorTime } = acc;
+    acc.accumulators[lastAccumulatorTime] = lastAccumulator;
+};
+
+const getTWAAccumulator = (acc, time) => {
+    return acc.accumulators[time];
+};
+
+const getTWA = (acc, start, end) => {
+    const { accumulators } = acc;
+    const startAccumulator = accumulators[start];
+    const endAccumulator = accumulators[end];
+    return endAccumulator.sub(startAccumulator).div(Decimal(end.toString()).sub(Decimal(start.toString())));
+};
+
+describe('TokenTimeWeightedAverage', () => {
+    const owner = defaultSender;
+    const seeder = accounts[1];
+    const nonOwner = accounts[5];
+    let twa;
+    let now;
+
+    beforeEach(async () => {
+        now = await latest();
```

Comment:
Since we're using `TestTokenTimeWeightedAverage`, which inherits from `TestTime`, my recommendation is to avoid using the actual blockchain time throughout the tests, as it just adds an undetermined factor into them (i.e., the possibility of different behavior on different executions).
You can just use any value larger than 0, or rely on the fact that `TestTime` already does this for you (that's why I've explicitly used `uint256 public currentTime = 1`), which will relieve you of the need to call `twa.setTime(now)`.

Reply:
My original intention was to mix "real" timestamps to the calculation, but since it's using a delta anyway - it doesn't matter. I'll fix it